### PR TITLE
Progress indication for File table engine in clickhouse-local and insert from file in clickhouse-client

### LINKED ISSUE #18209: Progress indication for File table engine.
To show progress bar we need to estimate the total number of rows.
We can do it by simple linear approximation if we know the file size, the amount of already processed rows and bytes.

**Use case**

It's very useful in `clickhouse-local` when large files are processed.

**Additional info**

When there are multiple files, we can also make some decent estimate.

We can also display a progress bar on INSERT query in clickhouse-client if it was invoked in the following form:

```
clickhouse-client --progress --query "INSERT ..." < file
```

### PR DESCRIPTION
**Pull request by kssenii**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Display progress for File table engine in clickhouse-local and on INSERT query in clickhouse-client when data is passed to stdin. Closes https://github.com/ClickHouse/ClickHouse/issues/18209.

### CODE REVIEW

**Comment by kssenii** [comment]
In my terminal it looks like this
Screenshot from 2021-04-27 21-52-29

**Comment by alexey-milovidov** [comment]
It looks like a custom color scheme. It is intended to be green but this "pony pink" color looks great nevertheless.

**Comment by kssenii** [comment]
btw, if there is `>`  in non-interactive mode in clickhouse-client, it results in `isatty(STDIN_FILENO) == false` (in `getTerminalWidth()`) => it cannot get terminal width and progress will be shown only with progress percentage and the amount of how much is already processed / left to process (same as on the picture for clickhouse-local but without running line).

**Comment by kssenii** [comment]
Will be finished after https://github.com/ClickHouse/ClickHouse/pull/23196.

**Comment by alexey-milovidov** [comment]
It is possible to check for terminal and obtain its properties with `/dev/tty` file.

**Comment by alexey-milovidov** [comment]
@kssenii Also a thought: as we output progress to stderr, we can check if stderr is a terminal and obtain it's width.

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by kssenii** [comment]
@Mergifyio update

**Comment by alexey-milovidov** [comment]
> 2021-06-09 10:07:18 ld.lld: error: src/libdbmsd.a: could not get the member for symbol DB::KeeperStorage::processRequest(std::__1::shared_ptr<Coordination::ZooKeeperRequest> const&, long, std::__1::optional<long>): truncated or malformed archive (terminator characters in archive member "'w" not the correct "`\n" values for the archive member header for O�'wO�'wO�'wO�'w)

It is either due to excessive binary size or if compiler job was terminated and file is truncated.

**Comment by alexey-milovidov** [comment]
@mergifyio update

File: src/Common/ProgressIndication.cpp [hunk 1]
```diff
@@ -1,22 +1,69 @@
-#include "ProgressBar.h"
+#include "ProgressIndication.h"
 #include <IO/WriteBufferFromFileDescriptor.h>
 #include <Common/TerminalSize.h>
 #include <Common/UnicodeBar.h>
 #include <Databases/DatabaseMemory.h>
 
+/// FIXME: progress bar in clickhouse-local needs to be cleared after query execution
```
Comment (akuzm) [inline src/Common/ProgressIndication.cpp:7]: Should this be fixed before we merge?

Comment (kssenii) [reply src/Common/ProgressIndication.cpp:7]: It should be done here https://github.com/ClickHouse/ClickHouse/pull/24535.

File: src/Common/ProgressIndication.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    /// In some cases there is a need to update progress value, when there is no access to progress_inidcation object.
+    /// In this case it is added via context.
+    /// `write_progress_on_update` is needed to write progress for loading files data via pipe in non-interactive mode.
+    void setFileProgressCallback(ContextMutablePtr context, bool write_progress_on_update = false);
+
+    /// How much seconds passed since query execution start.
+    UInt64 elapsedSeconds() const { return watch.elapsedSeconds(); }
+
+private:
+    /// Show progress bar only if on first writeProgress call the query is not yet half done.
```
Comment (akuzm) [inline src/Common/ProgressIndication.h:45]:
```suggestion
    /// This flag controls whether to show the progress bar. We start showing it after
    /// the query has been executing for 0.5 seconds, and is still less than half complete.
```

**Review by akuzm — APPROVED** [review]
LGTM overall.

**Comment by kssenii** [comment]
`Yandex synchronization check ` - `fatal error: 'clickhouse/base/ext/shared_ptr_helper.h' file not found` - not related

**Pull request merged**